### PR TITLE
Add support for building relocatable package and .deb/.rpm

### PR DESCRIPTION
**Pull request by fruch**
* copied all the needed machinary script from scylla-tools
* download dependecy from pypi, and repack them as zip files, so we can vendor them into deb/rpm/relocatable

### CODE REVIEW

File: dist/debian/control.template [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Source: %{product}-cqlsh
+Section: misc
+Priority: extra
+Maintainer: Takuya ASADA <syuu@scylladb.com>
```
Comment (avikivity) [inline dist/debian/control.template:4]: Copied too much

File: dist/debian/changelog.template [hunk 2]
```diff
@@ -0,0 +1,5 @@
+%{product}-cqlsh (%{version}-%{release}-%{revision}) %{codename}; urgency=medium
+
+  * New release 
+
+ -- Takuya ASADA <syuu@scylladb.com>  Mon, 24 Aug 2015 09:22:55 +0000
```
Comment (avikivity) [inline dist/debian/changelog.template:5]: Wrong name

File: dist/debian/control.template [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+Rules-Requires-Root: no
+
+Package: %{product}-cqlsh
+Architecture: all
+Depends: python3
+Recommends: ntp | time-daemon
+Conflicts: cassandra
+Description: Scylla cqlsh tool
+ Scylla is a highly scalable, eventually consistent, distributed,
+ partitioned row DB.
```
Comment (avikivity) [inline dist/debian/control.template:16]: Please adjust to explain that this is a shell for ScyllaDB.

File: dist/debian/debian/README.Debian [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Cassandra for Debian
+====================
+
+This package is not a part of Debian, (and there are no immediate plans
+to have it added). Bugs should be sent to the Apache Cassandra JIRA, *not*
+filed in the Debian BTS.
```
Comment (avikivity) [inline dist/debian/debian/README.Debian:6]: Looks unrelated.

File: dist/debian/debian/copyright [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+Upstream-Contact: http://www.scylladb.com/
+Source: https://github.com/scylladb/scylla-jmx
+
+Files: *
+Copyright: Copyright (c) 2009, The Apache Software Foundation.
+License: Apache-2.0
+
+Files: scripts/git-archive-all
+Copyright: Copyright (c) 2010 Ilya Kulakov
+License: MIT
```
Comment (avikivity) [inline dist/debian/debian/copyright:12]: Is it included?

File: dist/debian/debian_files_gen.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+# (at your option) any later version.
+#
+# Scylla is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+#
```
Comment (avikivity) [inline dist/debian/debian_files_gen.py:22]: cqlsh is not AGPL licensed, we should keep it under the original license.

File: dist/redhat/scylla-cqlsh.spec [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+%{_bindir}/cqlsh
+%{_bindir}/cqlsh.py
+/opt/scylladb/share/cassandra/bin/cqlsh
+/opt/scylladb/share/cassandra/bin/cqlsh.py
+/opt/scylladb/share/cassandra/pylib/*
+/opt/scylladb/share/cassandra/lib/*
+
+
+%changelog
+* Fri Aug  7 2015 Takuya ASADA Takuya ASADA <syuu@cloudius-systems.com>
```
Comment (avikivity) [inline dist/redhat/scylla-cqlsh.spec:40]: Bad copy

File: install.sh [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+# (at your option) any later version.
+#
+# Scylla is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+#
```
Comment (avikivity) [inline install.sh:21]: Wrong license

File: SCYLLA-VERSION-GEN [hunk 9]
```diff
@@ -0,0 +1,49 @@
+#!/bin/sh
+
+PRODUCT=scylla
+VERSION=666.development
+
+if test -f version
```
Comment (avikivity) [inline SCYLLA-VERSION-GEN:6]: Why is this file part of "enable integration tests"?

File: requirements.txt [hunk 10]
```diff
@@ -0,0 +1,6 @@
+scylla-driver==3.25.7
+six==1.16.0
+geomet==0.2.1.post1
+PyYAML==6.0
```
Comment (avikivity) [inline requirements.txt:4]: Do we need PyYAML for cqlsh? geomet?

**Comment by avikivity** [comment]
@syuu1228 please review

File: requirements.txt (hunk 10 above)
Comment (fruch) [reply requirements.txt:4]: yes, for the python-driver
and this list is the exact versions that gonna be bundled

File: dist/debian/control.template [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+Priority: extra
+Maintainer: Takuya ASADA <syuu@scylladb.com>
+Homepage: http://www.scylladb.com
+Standards-Version: 3.9.5
+Rules-Requires-Root: no
+
+Package: %{product}-cqlsh
+Architecture: all
+Depends: python3
+Recommends: ntp | time-daemon
```
[the remark this answers is not in the document]
Comment (fruch) [reply dist/debian/control.template:12]: removing

File: dist/debian/control.template (hunk 3 above)
Comment (fruch) [reply dist/debian/control.template:16]: is this good enough:
```
 scylla-cqlsh is tool to connect to Scylla or Cassandra clusters, for debugging 
 and dynamically do queries from a CLI 
```

File: dist/debian/debian/README.Debian (hunk 4 above)
Comment (fruch) [reply dist/debian/debian/README.Debian:6]: removing this file

File: dist/debian/debian/copyright (hunk 5 above)
Comment (fruch) [reply dist/debian/debian/copyright:12]: no, removing it

File: SCYLLA-VERSION-GEN (hunk 9 above)
Comment (fruch) [reply SCYLLA-VERSION-GEN:6]: it's a mistke

File: dist/debian/debian_files_gen.py (hunk 6 above)
Comment (fruch) [reply dist/debian/debian_files_gen.py:22]: replaced

File: dist/redhat/scylla-cqlsh.spec (hunk 7 above)
Comment (fruch) [reply dist/redhat/scylla-cqlsh.spec:40]: I think we never do edit the changelog on our deb/rpm, can they completed be removed ?

File: install.sh (hunk 8 above)
Comment (fruch) [reply install.sh:21]: replaced

File: SCYLLA-VERSION-GEN (hunk 9 above)
Comment (fruch) [reply SCYLLA-VERSION-GEN:6]: move into the correct commits

File: dist/debian/control.template (hunk 3 above)
Comment (avikivity) [reply dist/debian/control.template]: suggest s/do/perform/

File: dist/redhat/scylla-cqlsh.spec (hunk 7 above)
Comment (avikivity) [reply dist/redhat/scylla-cqlsh.spec]: I don't know, rpm-build may complain, or not.

Comment (fruch) [reply dist/redhat/scylla-cqlsh.spec]: It only does a warning if it's empty

File: dist/debian/debian/compat [hunk 12]
```diff
@@ -0,0 +1 @@
+9
```
Comment (yaronkaikov) [inline dist/debian/debian/compat:1]: What is this?

File: dist/debian/debian_files_gen.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    changelog_template = f.read()
+
+with open(os.path.join(scriptdir, 'control.template')) as f:
+    control_template = f.read()
+
+with open('build/SCYLLA-PRODUCT-FILE') as f:
+    product = f.read().strip()
+
+with open('build/SCYLLA-VERSION-FILE') as f:
+    version = f.read().strip().replace('.rc', '~rc').replace('_', '-')
```
Comment (yaronkaikov) [inline dist/debian/debian_files_gen.py:40]: This line is not needed anymore, since our current version syntax is `x.x.x-rcX` , it should also removed from Scylla

File: dist/debian/debian_files_gen.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+shutil.copytree('dist/debian/debian', 'build/debian/debian')
+
+if product != 'scylla':
+    for p in Path('build/debian/debian').glob('scylla-*'):
+        # pat1: scylla-server.service
+        #    -> scylla-enterprise-server.scylla-server.service
+        # pat2: scylla-server.scylla-fstrim.service
+        #    -> scylla-enterprise-server.scylla-fstrim.service
+        # pat3: scylla-conf.install
+        #    -> scylla-enterprise-conf.install
```
Comment (yaronkaikov) [inline dist/debian/debian_files_gen.py:56]: Do we need it?

File: dist/debian/debian_files_gen.py (hunk 13 above)
Comment (fruch) [reply dist/debian/debian_files_gen.py:40]: this file was copied as is from scylla-java-tools, I didn't change a thing here.

File: dist/debian/debian_files_gen.py (hunk 14 above)
Comment (fruch) [reply dist/debian/debian_files_gen.py:56]: this file was copied as is from scylla-java-tools, I didn't change a thing here.

File: dist/debian/debian/compat (hunk 12 above)
Comment (fruch) [reply dist/debian/debian/compat:1]: https://www.debian.org/doc/manuals/maint-guide/dother.en.html#compat

File: dist/debian/debian_files_gen.py (hunk 14 above)
Comment (avikivity) [reply dist/debian/debian_files_gen.py:56]: This isn't a good response. If it's wrong, let's not copy it. If it's right then ok.

File: dist/debian/debian_files_gen.py (hunk 13 above)
Comment (fruch) [reply dist/debian/debian_files_gen.py:40]: as if we are changing this on core, we could change it across the sub-modules.

and if I've got you correctly, we shouldn't be replacing anything here ?

Comment (yaronkaikov) [reply dist/debian/debian_files_gen.py:40]: yes, this condition is not relevant anymore. but since it's a new file. there is no point leaving it and then remove it IMO.

File: reloc/build_reloc.sh [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+if [ "$CLEAN" = "yes" ]; then
+    rm -rf build target
+fi
+
+if [ -f "$DEST" ]; then
+    rm "$DEST"
+fi
+
+if [ -z "$NODEPS" ]; then
+    sudo ./install-dependencies.sh
```
Comment (syuu1228) [inline reloc/build_reloc.sh:64]: install-dependencies.sh does not found.
We need following patch to fix:
https://github.com/syuu1228/scylla-cqlsh/commit/bf94cd92ecd17193c40e0294f00432596ac5ca06

File: install.sh [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+else
+    retc="$rprefix/$etcdir"
+fi
+
+install -d -m755 "$rprefix"/share/cassandra/bin
+if ! $nonroot; then
+    install -d -m755 "$rusr"/bin
+fi
+
+thunk="#!/usr/bin/env bash
```
Comment (syuu1228) [inline install.sh:91]: We should use relocatable python3 to run ALL our python script.
Also, we should create symlink on /usr/bin/cqlsh, instead of writing thunk to /usr/bin/cqlsh.
Because we want to use relocatable python3 both on /opt/scylladb/share/cassandra/bin/cqlsh and /usr/bin/cqlsh.


We need following fix:
https://github.com/syuu1228/scylla-cqlsh/commit/4ba3c7a4a2e31a9a13903e8ea9a4ac63459a8c67

**Review by syuu1228 — CHANGES_REQUESTED** [review]

File: dist/debian/debian_files_gen.py (hunk 14 above)
Comment (syuu1228) [reply dist/debian/debian_files_gen.py:56]: We need to keep this just for scylla-cqlsh.install.
And we actually drop scylla-cqlsh.install, since our .deb build script is using .install file for producing mutiple .deb package (sub-package) on single repository.
We install all files to $(CURDIR)/debian/tmp and then picking files by *.install file for each sub-packages.
Since scylla-cqlsh is single package, we just need to install $(CURDIR)/debian/scylla-cqlsh and stop using .install file.

The fix will be like this:
https://github.com/syuu1228/scylla-cqlsh/commit/eeeb1fae24cb9ef63c6c531a89bedf57c2744226

File: dist/debian/control.template [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+Maintainer: Israel Fruchter <fruch@scylladb.com>
+Homepage: http://www.scylladb.com
+Standards-Version: 3.9.5
+Rules-Requires-Root: no
+
+Package: %{product}-cqlsh
+Architecture: all
+Depends: python3
+Conflicts: cassandra
+Description: cqlsh is a Python-based command-line client for running CQL commands on a cassandra cluster.
```
Comment (syuu1228) [inline dist/debian/control.template:13]: When we splited scylla-tools package into two (scylla-tools and scylla-tools-core), we added Replaces and Breaks on debian/control to avoid conflict while upgrading packages, like this:
https://github.com/scylladb/scylla-tools-java/commit/59531395df77dc19da5638df336a1e8dcadf0d6c
(It fixes scylladb/scylladb#2732 and scylladb/scylladb#2765)

I guess we need similar one here, something like:
```
Replaces: scylla-tools (<< 5.2~rc0)
Breaks: scylla-tools (<< 5.2~rc0)
```

Note that I haven't tested this yet, we need to verify upgrade works with scylla-cqlsh, and make sure it's not conflict with older scylla-tools.

Comment (fruch) [reply dist/debian/control.template:13]: add this change, and tested it, it's works as expected (installed 5.1 scylla-tools, and then scylla-cqlsh)

File: install.sh (hunk 16 above)
Comment (fruch) [reply install.sh]: took this change, but still this package isn't depended on the relocatable python3, can be installed on it's own and would work without relocatable python3

File: reloc/build_reloc.sh (hunk 15 above)
Comment (fruch) [reply reloc/build_reloc.sh:64]: took this patch too

File: dist/debian/debian_files_gen.py (hunk 14 above)
Comment (fruch) [reply dist/debian/debian_files_gen.py]: took the suggested patch

File: dist/debian/debian_files_gen.py (hunk 13 above)
Comment (fruch) [reply dist/debian/debian_files_gen.py]: removed this hack

**Comment by fruch** [comment]
all of @syuu1228 comments were addressed, and those changed passed in scylla-CI

**Review by yaronkaikov — APPROVED** [review]
